Model training encounters error nondeterministically

**What is the current behavior?**
When training a model, sometimes the following error is encountered.
What is strange is that I am training two identical models, on identical data (differentiated using 2 different workspace names), and the error is encountered for one of them only (although on some rare occasions, the error is encountered for both).

```
[orchestrator_api.py:390] starting iteration 0 in background for workspace 'WS1' category id '0' using 7 items
[orchestrator_api.py:602] Workspace 'WS2' category id '0' 4 positive elements (>=4) 7 elements changed since last model (>=4). Training a new model
[orchestrator_api.py:398] workspace 'WS1' training a model for category id '0', train_statistics: {'train_counts': {'False': 3, 'True': 4}}
[train_set_selectors.py:38] using 7 for train using dataset DATASET
[orchestrator_api.py:390] starting iteration 0 in background for workspace 'WS2' category id '0' using 7 items
[orchestrator_api.py:398] workspace 'WS2' training a model for category id '0', train_statistics: {'train_counts': {'False': 3, 'True': 4}}
[models_background_jobs_manager.py:37] Adding training for model id SVM_BOW_289e6de2-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
[models_background_jobs_manager.py:37] Adding training for model id SVM_BOW_289e8d0e-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
[models_background_jobs_manager.py:37] Adding training for model id SVM_GloVe_289e6de3-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
[models_background_jobs_manager.py:37] Adding training for model id SVM_GloVe_289e8d0f-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
[tools.py:67] Done getting GloVe representations for 7 sentences
[tools.py:67] Done getting GloVe representations for 7 sentences
Adding training for model id SVM_BOW_289e6de2-1a94-11ed-add3-296780c3e733,SVM_GloVe_289e6de3-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
training an ensemble model id SVM_BOW_289e6de2-1a94-11ed-add3-296780c3e733,SVM_GloVe_289e6de3-1a94-11ed-add3-296780c3e733 using 7 elements
[models_background_jobs_manager.py:37] Adding training for model id SVM_BOW_289e8d0e-1a94-11ed-add3-296780c3e733,SVM_GloVe_289e8d0f-1a94-11ed-add3-296780c3e733 into the CPU_10_threadpool
[ensemble.py:77] training an ensemble model id SVM_BOW_289e8d0e-1a94-11ed-add3-296780c3e733,SVM_GloVe_289e8d0f-1a94-11ed-add3-296780c3e733 using 7 elements
ERROR    [_base.py:332] exception calling callback for <Future at 0x7f0195972ca0 state=finished returned str>
Traceback (most recent call last):
  File "/home/farnaz/anaconda3/envs/label-sleuth/lib/python3.9/concurrent/futures/_base.py", line 330, in _invoke_callbacks
    callback(self)
  File "/root/label-sleuth/label_sleuth/orchestrator/orchestrator_api.py", line 433, in _train_done_callback
    self.orchestrator_state.update_model_status(workspace_id=workspace_id, category_id=category_id,
  File "/root/label-sleuth/label_sleuth/orchestrator/core/state_api/orchestrator_state_api.py", line 278, in update_model_status
    assert len(iterations) > iteration_index,\
AssertionError: Iteration '0' doesn't exist in workspace 'WS2'
```


Note that iteration '0' is not the only iteration where this can happen. In one case I encountered this error in iteration '2', with 15 labeled datapoints:
`train_statistics: {'train_counts': {'True': 9, 'False': 6}}`


From then on, the logs show the following message for the workspace:
`workspace 'WS2' category id '0' new elements criterion was met but previous AL not yet ready, not initiating a new training`

**Steps to reproduce the problem:**
These settings are changed from the default config:
```
"first_model_positive_threshold": 4,
"changed_element_threshold": 4,
"training_set_selection_strategy": "ALL_LABELED"
```

This error is encountered non-deterministically (not always encountered).

**Environment:**
 - OS: Ubuntu 20.04.4 LTS
 - Python version: 3.9.12

--- a/label_sleuth/orchestrator/orchestrator_api.py
+++ b/label_sleuth/orchestrator/orchestrator_api.py
@@ -198,9 +198,9 @@ class OrchestratorApi:
         :param remove_duplicates:
         :return: a list of TextElement objects
         """
-        return self.data_access.get_labeled_text_elements(workspace_id, dataset_name, category_id,
-                                                          sample_size=sys.maxsize, remove_duplicates=remove_duplicates)[
-            'results']
+        return self.data_access.get_labeled_text_elements(
+            workspace_id, dataset_name, category_id, sample_size=sys.maxsize,
+            remove_duplicates=remove_duplicates)['results']
 
     def get_all_unlabeled_text_elements(self, workspace_id, dataset_name, category_id: int, remove_duplicates=False) \
             -> List[TextElement]:
@@ -398,15 +398,15 @@ class OrchestratorApi:
 
         logging.info(f"workspace '{workspace_id}' training a model for category id '{category_id}', "
                      f"train_statistics: {train_statistics}")
-        model_id, _ = model.train(train_data=train_data, language=self.config.language,
-                                  done_callback=functools.partial(self._train_done_callback, workspace_id,
-                                                                  category_id, new_iteration_index))
+        model_id, future = model.train(train_data=train_data, language=self.config.language)
         model_status = model.get_model_status(model_id)
         model_info = ModelInfo(model_id=model_id, model_status=model_status, model_type=model_type,
                                train_statistics=train_statistics, creation_date=datetime.now())
         self.orchestrator_state.add_iteration(workspace_id=workspace_id, category_id=category_id,
                                               model_info=model_info)
-
+        # The train callback is added here to ensure it only runs after the iteration has been added
+        future.add_done_callback(functools.partial(self._train_done_callback, workspace_id, category_id,
+                                                   new_iteration_index))
         # The model id is returned almost immediately, but the training is performed in the background. Once training is
         # complete the iteration flow continues in the *_train_done_callback* method
         return model_id

